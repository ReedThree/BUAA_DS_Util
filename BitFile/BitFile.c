#include "BitFile.h"
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

struct BitFile *BitFile_create(FILE *file) {
    struct BitFile *bfile = (struct BitFile *)malloc(sizeof(struct BitFile));
    bfile->inner = file;
    bfile->bitPos = 0;
    bfile->buffer = 0;
    bfile->bufferLen = 0;
    return bfile;
}

void BitFile_flush(struct BitFile *bfile) {
    if (bfile->bufferLen > 0) {
        BitFile_padding(bfile);
        fputc(bfile->buffer, bfile->inner);
    }
}
void BitFile_close(struct BitFile *bfile) {
    BitFile_flush(bfile);
    free(bfile);
}

size_t BitFile_readBits(uint8_t *buffer, size_t len, struct BitFile *bfile) {
    size_t bytesToRead = len / 8 + 2;

    uint8_t *bytesBuffer = (uint8_t *)malloc(bytesToRead * sizeof(uint8_t));
    size_t innerBeginPos = bfile->bitPos / 8;

    if (fseek(bfile->inner, (long)innerBeginPos, SEEK_SET) != 0) {
        return 0;
    }

    size_t successBytesCount =
        fread(bytesBuffer, sizeof(uint8_t), bytesToRead, bfile->inner);

    if (successBytesCount == 0) {
        return 0;
    }

    size_t bufferIndex = 0;

    if (bfile->bitPos % 8 != 0) {
        uint8_t leadingBitsCount =
            (len >= 8) ? (uint8_t)(8 - (bfile->bitPos % 8)) : (uint8_t)len;

        for (uint8_t i = 0; i < leadingBitsCount; i++) {
            buffer[bufferIndex] =
                (bytesBuffer[0] >> (leadingBitsCount - i - 1)) & 1;
            bufferIndex++;
        }
    } else {
        uint8_t leadingBitsCount =
            (len >= 8) ? (uint8_t)(8 - (bfile->bitPos % 8)) : (uint8_t)len;

        for (uint8_t i = 0; i < leadingBitsCount; i++) {
            buffer[bufferIndex] = (bytesBuffer[0] >> (7 - i)) & 1;
            bufferIndex++;
        }
    }

    for (size_t i = 1; i < successBytesCount - 1; i++) {
        for (uint8_t j = 0; j < 8; j++) {
            buffer[bufferIndex] = (bytesBuffer[i] >> (7 - j)) & 1;
            bufferIndex++;
        }
    }

    if (successBytesCount == bytesToRead) {
        uint8_t endingBitsCount = (bfile->bitPos + len + 1) % 8;
        for (uint8_t i = 0; i < endingBitsCount; i++) {
            if (bufferIndex >= len) {
                break;
            }
            buffer[bufferIndex] =
                (bytesBuffer[successBytesCount - 1] >> (7 - i)) & 1;
            bufferIndex++;
        }
    } else {
        if (successBytesCount > 1) {
            uint8_t endingBitsCount = (uint8_t)(len - bufferIndex);
            endingBitsCount = (endingBitsCount >= 8) ? 8 : endingBitsCount;
            for (uint8_t i = 0; i < endingBitsCount; i++) {
                buffer[bufferIndex] =
                    (bytesBuffer[successBytesCount - 1] >> (7 - i)) & 1;
                bufferIndex++;
            }
        }
    }

    bfile->bitPos += bufferIndex;
    free(bytesBuffer);
    return bufferIndex;
}
size_t BitFile_writeBits(uint8_t *buffer, size_t len, struct BitFile *bfile) {
    fseek(bfile->inner, (long)bfile->bitPos / 8, SEEK_SET);
    size_t bufferIndex = 0;
    if (bfile->bitPos % 8 != 0) {
        int ch = fgetc(bfile->inner);
        if (ch == EOF) {
            return 0;
        }

        uint8_t firstByte = (uint8_t)ch;
        uint8_t firstWriteBitsCount = (uint8_t)(8 - (bfile->bitPos % 8));

        for (uint8_t i = 0; i < firstWriteBitsCount; i++) {
            firstByte =
                (uint8_t)((firstByte & ~(1 << (firstWriteBitsCount - i - 1))) |
                          (buffer[bufferIndex]
                           << (firstWriteBitsCount - i - 1)));
            bufferIndex++;
        }

        fseek(bfile->inner, (long)bfile->bitPos / 8, SEEK_SET);
        fputc(firstByte, bfile->inner);
    }

    uint8_t currentByte = 0;
    uint8_t currentByteLen = 0;

    while (bufferIndex < len) {
        currentByte = (uint8_t)((currentByte << 1) | buffer[bufferIndex]);
        currentByteLen++;

        if (currentByteLen == 8) {
            fputc(currentByte, bfile->inner);
            currentByteLen = 0;
        }

        bufferIndex++;
    }

    if (currentByteLen != 0) {
        currentByte = (uint8_t)(currentByte << (8 - currentByteLen));
        fputc(currentByte, bfile->inner);
        bufferIndex += 8;
    }

    bfile->bitPos += len;
    return bufferIndex;
}

void BitFile_padding(struct BitFile *bfile) {
    if (bfile->bitPos % 8 != 0) {
        uint8_t paddingCount = (uint8_t)(8 - (bfile->bitPos % 8));
        bfile->buffer = (uint8_t)(bfile->buffer << paddingCount);
        bfile->bitPos += paddingCount;
    }
}

void BitFile_seek(struct BitFile *bfile, size_t bitPos) {
    bfile->bitPos = bitPos;
}
