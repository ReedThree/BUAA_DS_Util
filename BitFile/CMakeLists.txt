add_library(BitFile BitFile.c)
target_link_libraries(BitFile PRIVATE util_compiler_flags)
target_include_directories(BitFile INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
