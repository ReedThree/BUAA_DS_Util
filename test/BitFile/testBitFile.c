#include "BitFile.h"
#include <stdint.h>
#include <stdio.h>
#include <string.h>

int main(void) {
    FILE *f = fopen("write_5bits.bin", "w");

    uint8_t bits[] = {1, 0, 0, 1, 1};
    struct BitFile *bf = BitFile_create(f);
    BitFile_writeBits(bits, 5, bf);
    printf("1: bitPos: %zu\n", bf->bitPos);
    BitFile_close(bf);

    fclose(f);

    f = fopen("write_5bits.bin", "r");
    bf = BitFile_create(f);
    uint8_t bits2[5] = {0};
    BitFile_readBits(bits2, 5, bf);
    printf("2:\n");
    for (size_t i = 0; i < 5; i++) {
        printf("%hhu ", bits2[i]);
    }
    putchar('\n');
    printf("3: bitPos: %zu\n", bf->bitPos);
    BitFile_close(bf);

    fclose(f);

    f = fopen("write_17bits.bin", "w");
    bf = BitFile_create(f);
    // 0x9C 0x9C + 1
    uint8_t bits3[17] = {1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1};

    BitFile_writeBits(bits3, 17, bf);
    printf("4: bitPos: %zu\n", bf->bitPos);
    BitFile_close(bf);

    fclose(f);

    f = fopen("write_17bits.bin", "r");
    bf = BitFile_create(f);
    uint8_t bits4[17] = {0};

    BitFile_readBits(bits4, 17, bf);
    printf("5:\n");
    for (size_t i = 0; i < 17; i++) {
        printf("%hhu ", bits4[i]);
    }
    putchar('\n');
    printf("6: bitPos: %zu\n", bf->bitPos);
    BitFile_close(bf);

    fclose(f);

    f = fopen("write_17bits_seek_5bits.bin", "w");
    bf = BitFile_create(f);
    BitFile_writeBits(bits3, 17, bf);
    BitFile_close(bf);
    fclose(f);

    f = fopen("write_17bits_seek_5bits.bin", "r+");
    bf = BitFile_create(f);
    BitFile_seek(bf, 5);
    BitFile_writeBits(bits3, 17, bf);
    BitFile_close(bf);
    fclose(f);

    f = fopen("write_17bits_seek_5bits.bin", "r");
    bf = BitFile_create(f);

    uint8_t bits5[22] = {0};
    BitFile_readBits(bits5, 22, bf);

    printf("7:\n");
    for (size_t i = 0; i < 22; i++) {
        printf("%hhu ", bits5[i]);
    }
    putchar('\n');

    BitFile_close(bf);
    fclose(f);

    f = fopen("write_17bits_seek_5bits.bin", "r");
    bf = BitFile_create(f);
    uint8_t bits6[17] = {0};
    BitFile_seek(bf, 5);
    BitFile_readBits(bits6, 17, bf);

    printf("8:\n");
    for (size_t i = 0; i < 17; i++) {
        printf("%hhu ", bits6[i]);
    }
    putchar('\n');

    BitFile_seek(bf, 17);
    memset(bits6, 0, 17);
    printf("9: actualReadBits: %zu\n", BitFile_readBits(bits6, 17, bf));

    printf("10:\n");
    for (size_t i = 0; i < 17; i++) {
        printf("%hhu ", bits6[i]);
    }
    putchar('\n');

    BitFile_close(bf);
    fclose(f);

    f = fopen("write_twice_5bits.bin", "w");
    bf = BitFile_create(f);

    uint8_t bits7[5] = {1, 0, 0, 1, 1};
    BitFile_writeBits(bits7, 5, bf);
    BitFile_writeBits(bits7, 5, bf);
    BitFile_close(bf);
    fclose(f);

    f = fopen("write_twice_5bits.bin", "r");
    bf = BitFile_create(f);

    uint8_t bits8[5] = {0};
    BitFile_readBits(bits8, 5, bf);

    printf("11:\n");
    for (size_t i = 0; i < 5; i++) {
        printf("%hhu ", bits8[i]);
    }
    putchar('\n');

    memset(bits8, 0, 5);
    BitFile_readBits(bits8, 5, bf);
    printf("12:\n");
    for (size_t i = 0; i < 5; i++) {
        printf("%hhu ", bits8[i]);
    }
    putchar('\n');

    BitFile_close(bf);
    fclose(f);

    return 0;
}