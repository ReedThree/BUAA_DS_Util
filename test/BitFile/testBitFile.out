1: bitPos: 5
2:
1 0 0 1 1 
3: bitPos: 5
4: bitPos: 17
5:
1 0 0 1 1 1 0 0 1 0 0 1 1 1 0 0 1 
6: bitPos: 17
7:
1 0 0 1 1 1 0 0 1 1 1 0 0 1 0 0 1 1 1 0 0 1 
8:
1 0 0 1 1 1 0 0 1 0 0 1 1 1 0 0 1 
9: actualReadBits: 7
10:
1 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 
11:
1 0 0 1 1 
12:
1 0 0 1 1 
